### ISSUE TITLE
BUG: using LinearOperator as RHS operand of @ causes a NumPy error

### ISSUE THREAD

**Issue by andportnoy** [scipy/scipy#17281]
### Describe your issue.

Using a `LinearOperator` on the right hand side of a matmul (through the `@` operator) leads to a NumPy error (see attached).

It does work as a left hand side operand. Moreover, `A @ B`, where `A` is a NumPy array and `B` is a `LinearOperator`, can be implemented using this expression as a workaround:
```python
(B.T @ A.T).T
```
A derived `LinearOperator` is thus forced to the left hand side.

It is not clear whether `A @ B` should be expected to work, so I'm creating this issue to start a discussion. See https://github.com/scikit-learn/scikit-learn/issues/12794#issuecomment-1287827481 for more context.

Fundamentally `A @ B` is equivalent to `A.__matmul__(B)`, so the implementation of `A` controls execution. To my knowledge as part of the general `ufunc` execution flow NumPy will try to coerce the `ufunc` arguments to `PyObjectArray`'s which are expected to have explicit data arrays. However as an abstraction of a 2D array, `LinearOperator`'s do not necessarily have any data. At the moment they end up being converted to zero-dimensional scalars, which leads to the error in question.

One could always materialize the data by letting the `LinearOperator` act on an identity matrix, but that is typically undesirable in matrix-free methods where this class is used.

As @ogrisel [notes](https://github.com/scikit-learn/scikit-learn/pull/18689#issuecomment-1287897926) though, it would be good to at least have a more user-friendly message.

### Reproducing Code Example

```python
import numpy as np
from scipy.sparse.linalg import LinearOperator

def linear_operator_from_array(A):
     return LinearOperator(
        shape=A.shape,
        dtype=A.dtype,
        matvec=lambda x: A @ x,
        rmatvec=lambda x: A.T @ x,
        matmat=lambda X: A @ X,
        rmatmat=lambda X: A.T @ X,
    )

A = np.empty((2, 3))
B = np.empty((3, 4))
B = linear_operator_from_array(B)
A @ B
```

### Error message

```shell
Traceback (most recent call last):
  File "/home/andrey/Projects/scikit-learn/linear-operator-matmul-fail.py", line 17, in <module>
    A @ B
ValueError: matmul: Input operand 1 does not have enough dimensions (has 0, gufunc core with signature (n?,k),(k,m?)->(n?,m?) requires 1)
```

### SciPy/NumPy/Python version information

1.9.1 1.23.4 sys.version_info(major=3, minor=10, micro=7, releaselevel='final', serial=0)

**Comment by andportnoy** [comment]
`np.array([]) @ None` leads to the same error. Maybe NumPy could be more conservative during `ufunc` argument conversion and reject certain inputs early.

**Comment by akhmerov** [comment]
I've implemented a rather uncontroversial fix in #18601.
